File: backend/src/models/Booking.js
const mongoose = require('mongoose');

const bookingSchema = new mongoose.Schema(
  {
    customer: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
    provider: { type: mongoose.Schema.Types.ObjectId, ref: 'Provider', required: true },
    bookingType: {
      type: String,
      enum: ['hourly', 'half_day', 'full_day', 'multi_day'],
      required: true,
    },
    eventType: { type: String, required: true },
    scheduledDate: { type: Date, required: true },
    startTime: { type: String, required: true },
    durationHours: Number,
    durationDays: Number,
    address: {
      line1: String,
      line2: String,
      city: String,
      state: String,
      district: String,
      pincode: String,
      coordinates: { lat: Number, lng: Number },
    },
    specialInstructions: String,
    amount: { type: Number, required: true },
    advanceAmount: { type: Number, required: true },
    remainingAmount: { type: Number, default: 0 },
    advancePercent: { type: Number, default: 15 },
    status: {
      type: String,
      enum: [
        'pending_payment',
        'confirmed',
        'in_progress',
        'awaiting_otp_verification',
        'ready_for_remaining_payment',
        'completed',
        'cancelled',
        'refunded',
      ],
      default: 'pending_payment',
    },
    payment: {
      advanceOrderId: String,
      advancePaymentId: String,
      advancePaid: { type: Boolean, default: false },
      remainingOrderId: String,
      remainingPaymentId: String,
      remainingPaid: { type: Boolean, default: false },
      remainingPaymentUnlocked: { type: Boolean, default: false },
      receiptNumber: String,
    },
    completionOtp: {
      code: String,
      expiresAt: Date,
      generatedAt: Date,
      customerVerified: { type: Boolean, default: false },
      providerVerified: { type: Boolean, default: false },
      customerVerifiedAt: Date,
      providerVerifiedAt: Date,
    },
    cancelledBy: { type: String, enum: ['customer', 'provider', 'admin'] },
    cancellationReason: String,
  },
  { timestamps: true }
);

bookingSchema.index({ customer: 1, createdAt: -1 });
bookingSchema.index({ provider: 1, scheduledDate: 1 });
bookingSchema.index({ status: 1 });

module.exports = mongoose.model('Booking', bookingSchema);

File: backend/src/main/java/com/pujasetu/model/Booking.java
package com.pujasetu.model;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.pujasetu.model.enums.BookingStatus;
import com.pujasetu.model.enums.BookingType;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.mapping.Field;

import java.time.Instant;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Document(collection = "bookings")
public class Booking {

    @Id
    @JsonProperty("_id")
    private String id;

    @Indexed
    @Field("customer")
    @JsonProperty("customer")
    private String customerId;

    @Indexed
    @Field("provider")
    @JsonProperty("provider")
    private String providerId;

    private BookingType bookingType;
    private String eventType;
    private Instant scheduledDate;
    private String startTime;
    private Integer durationHours;
    private Integer durationDays;
    private BookingAddress address;
    private String specialInstructions;

    private int amount;
    private int advanceAmount;
    private int remainingAmount;

    @Builder.Default
    private int advancePercent = 15;

    @Builder.Default
    private BookingStatus status = BookingStatus.PENDING_PAYMENT;

    private PaymentInfo payment;
    private CompletionOtp completionOtp;
    private String cancelledBy;
    private String cancellationReason;

    @CreatedDate
    private Instant createdAt;

    @LastModifiedDate
    private Instant updatedAt;

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class BookingAddress {
        private String line1;
        private String line2;
        private String city;
        private String state;
        private String district;
        private String pincode;
        private Coordinates coordinates;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class Coordinates {
        private Double lat;
        private Double lng;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class PaymentInfo {
        private String advanceOrderId;
        private String advancePaymentId;
        private boolean advancePaid = false;
        private String remainingOrderId;
        private String remainingPaymentId;
        private boolean remainingPaid = false;
        private boolean remainingPaymentUnlocked = false;
        private String receiptNumber;
    }

    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class CompletionOtp {
        private String code;
        private Instant expiresAt;
        private Instant generatedAt;
        private boolean customerVerified = false;
        private boolean providerVerified = false;
        private Instant customerVerifiedAt;
        private Instant providerVerifiedAt;
    }
}
